3.next - Command collection autoDiscover()

markstory opened the pull request:
Build out the `autoDiscover()` method for `CommandCollection` this method will scan the core, app and all plugins for shell commands and append them into the collection. I've made the results of the scan include enough data that I can refactor the `CommandListShell` and `CompletionShell` in a subsequent pull request.

Refs #10716

markstory committed:
Fix incorrect class name.

This class was missing the correct suffix.

markstory committed:
Implement command auto-discovery.

Auto-discovery will allow us to maintain backwards compatibility and
provide convention based defaults that people can replace with their own
behavior if they so wish.

markstory committed:
Fix up test that started failing because of new shell class.

markstory committed:
Always add long alias for plugin shells.

Always add the long alias for plugin shells for backwards compatibility
reasons. The ShellDispatcher implementatino would allow plugin shells to
either be called with their long name or short name if there were no
conflicts. Because we're building the list ahead of dispatch time we
need to include all the aliases now.

Bump up the logging as not being able to load classes is something
people will want to know.

markstory committed:
Use methods that exist.

lorenzo commented on src/Console/CommandCollection.php:
```diff
@@ -130,4 +133,62 @@ public function count()
+     *
+     * @return $this
+     */
+    public function autoDiscover()
```
What if autodiscover returned the list of commands instead. That would make the api look like

```php
$collection->addAll($collection->autoDiscover());
```

The benefit I see from doing this is that you could filter-out or reorder some of the results if needed.

markstory replied:
Good idea. I'll make that change.

markstory committed:
Make autoDiscover() return an array.

Returning an array of information allows users to manipulate the
discovered commands before adding the commands to the collection.

markstory committed:
Remove unused import.

ravage84 commented on src/Console/CommandCollection.php:
```diff
@@ -60,8 +61,9 @@ public function add($name, $command)
+                "Cannot use '$class' for command '$name' it is not a subclass of Cake\Console\Shell."
             );
         }
         $this->commands[$name] = $command;
```
Add white line as it has nothing to do with the check above.

ravage84 commented on src/Console/CommandCollection.php:
```diff
@@ -130,4 +148,60 @@ public function count()
+            if (empty($shells[$key])) {
+                return $out;
+            }
+            foreach ($shells[$key] as $info) {
```
Add white lline for readability

ravage84 commented on src/Console/CommandCollection.php:
```diff
@@ -130,4 +148,60 @@ public function count()
+        $scanner = new CommandScanner();
+        $shells = $scanner->scanAll();
+
+        $adder = function ($out, $shells, $key) {
```
Why not as a protected method?
I think this method could be shorter, mor readable, may be even more self-explanatory

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+ * Used by CommanCollection and CommandTask to scan the filesystem
+ * for command classes.
+ *
+ * @internal
```
👍

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+class CommandScanner
+{
+    /**
+     * Scan the application, cakephp and plugins for shells
```
CakePHP

ravage84 replied:
The order is mentiond differently, [above](https://github.com/cakephp/cakephp/pull/10767/files#diff-c4fae5f19f4310651de5ab7f45abcebfR158).

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+            '',
+            ['command_list']
+        );
+        $plugins = [];
```
Add a white line above for clearer segregation

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+        $shellList = [];
+
+        $appNamespace = Configure::read('App.namespace');
+        $shellList['app'] = $this->scanDir(
```
extract as ``scanApp()``? Could improve reusability.

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+            ['app']
+        );
+
+        $shellList['CORE'] = $this->scanDir(
```
extract as ``scanCore()``? Could improve reusability.

markstory replied:
But its a protected method on an internal class. How much reuse could there be?

ravage84 replied:
Well, might not be that relevant, would decrease method size and increase readabilty, anyway

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+            ['command_list']
+        );
+        $plugins = [];
+        foreach (Plugin::loaded() as $plugin) {
```
extract as ``scanPlugins()``? Could improve reusability.

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+        if (empty($contents[1])) {
+            return [];
+        }
+        $shells = [];
```
Add white line above for readability.

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+            if (substr($file, -4) !== '.php') {
+                continue;
+            }
+            $shell = substr($file, 0, -4);
```
Add white line above for readability.

ravage84 commented on src/Console/CommandScanner.php:
```diff
@@ -0,0 +1,105 @@
+            if (in_array($name, $hide, true)) {
+                continue;
+            }
+            $shells[] = [
```
Add white line above for readability.

markstory committed:
Improve readability.

lorenzo commented:
This looks good to me, great job!

markstory commented:
Thanks @lorenzo
